rc/src/lib.rs

```diff
@@ -0,0 +1,96 @@
+//MyRc
+
+use std::ops::Deref;
+use std::mem;
+
+pub struct MyRc<T>
+{
+	ptr: *mut Counter<T>,
+}
+
+pub struct Counter<T>
+{
+	data: T,
+	count: usize,
+}
+
+impl<T> MyRc<T>
+{
+	pub fn new(t: T) -> MyRc<T>
+	{
+		let c = Box::new(Counter{data: t, count: 1});
+		MyRc{ptr: Box::into_raw(c)}
+	}
+
+	pub fn consume(self) -> Result<T, MyRc<T>>
+	{
+		unsafe
+		{
+			if (*self.ptr).count == 1
+			{
+				//we're about to drop our only pointer to the data
+				//so it's ok to zero it out
+				let fake: Counter<T> = mem::uninitialized();
+
+				//get the real value by replacing it with fake
+				let real = std::ptr::replace(self.ptr, fake); 
+
+				//doesn't drop real.data
+				Ok(real.data)
+			}
```

Comment:
Unless I'm reading this wrong, after this block the destructor for `self` is going to run. When it's pointer is pointing to uninitialized data :/